care/facility/api/viewsets/asset.py

```diff
@@ -205,15 +208,28 @@ def retrieve(self, request, *args, **kwargs):
         return Response(hit)
 
 
-class AssetAvailabilityFilter(filters.FilterSet):
-    external_id = filters.CharFilter(field_name="asset__external_id")
+class AssetAvailabilityViewSet(ListModelMixin, RetrieveModelMixin, GenericViewSet):
+    queryset = AvailabilityRecord.objects.all()
+    serializer_class = AvailabilityRecordSerializer
 
+    def get_queryset(self):
+        return self.queryset.filter(
+            object_external_id=self.kwargs["asset_external_id"],
+            content_type__model="asset",
+        )
 
-class AssetAvailabilityViewSet(ListModelMixin, RetrieveModelMixin, GenericViewSet):
-    queryset = AssetAvailabilityRecord.objects.all().select_related("asset")
-    serializer_class = AssetAvailabilitySerializer
-    filter_backends = (filters.DjangoFilterBackend,)
-    filterset_class = AssetAvailabilityFilter
+
+class AssetLocationAvailabilityViewSet(
+    ListModelMixin, RetrieveModelMixin, GenericViewSet
+):
+    queryset = AvailabilityRecord.objects.all()
+    serializer_class = AvailabilityRecordSerializer
+
+    def get_queryset(self):
+        return self.queryset.filter(
+            object_external_id=self.kwargs["asset_location_external_id"],
+            content_type__model="assetlocation",
+        )
```

Comment:
needs access checks

you can use a common viewset for both

add permission classes

and for access control call get_facility_queryset and then try to resolve the object using the facility queryset

Reply:
Changes made and tested. Also added tests for the permission checks.